### ISSUE TITLE
MQTT state machine doesn't detect disconnect

### ISSUE THREAD

**Issue by sgacond** [Azure/azure-iot-sdk-node#189]
# Context
- **OS and version used:** Raspbian Jessie, Linux 4.9.40-v7+, armv7l
- **Node.js version:** v9.3.0
- **npm version:** v5.5.1
- **list of installed packages:** 
```
├── azure-iot-common@1.3.0
├── azure-iot-device@1.3.0
├── azure-iot-device-mqtt@1.3.0
└── bluebird@3.5.1
```

# Description of the issue:
We used to have strange behaviour on some devices out in the field. 
The logs looked like the transmission timer suddenly stopped to transmit, but there was no exception or error callback. Devices with weaker and more unstable internet connection did it more often then others (some once a month, some once a week).
A simple restart of the whole app solved the issue for a while.

We found out, that losing the network connection and reconnecting it in on the raspbian device caused the iot-sdk to stop doing anything. Every method call (even "disconnect") went to neverland. No callback, no error. The network interface was not disconnected, just the cable. On my development machine (mac os), the behaviour was different: after a timeout time, the error callback arrived and the next "sendEvent" tried to reconnect.

The retry policy is set to NoRetry().

# Investigation

Unfortunately our devices don't have any display port and remote debugging while unplugging the network connection is a bit a challange. But using a second network adapter and a local connection without gateway address enabled that.

Unlike on my dev mac, the underlying socket under the MQTT lib was never closed on raspbian ("end-of-stream" was never fired). So the state machine of "MqttBase" didn't catch the "close"-event ([transport/mqtt/src/mqtt_base.ts:77](https://github.com/Azure/azure-iot-sdk-node/blob/55dd09dd2ba0d1bc331f9b4755d9c8a37399da99/common/transport/mqtt/src/mqtt_base.ts#L77)) of _mqttClient and stayed in connected state.

## Ugly workaround

As a workaround I introduced a timeout on the send event (bluebird: prmoisified and timeout(...)), and shifted state manually when it got hit.

# Code sample exhibiting the issue:
```javascript
const Message = require('azure-iot-device').Message,
      Client = require('azure-iot-device').Client,
      SharedAccessKeyAuthenticationProvider = require('azure-iot-device').SharedAccessKeyAuthenticationProvider,
      Mqtt = require('azure-iot-device-mqtt').Mqtt,
      NoRetry = require('azure-iot-common').NoRetry,
      Promise = require('bluebird');

let data = '{...}';

let authProvider = new SharedAccessKeyAuthenticationProvider({ 
    host: '...',
    deviceId: '...',
    sharedAccessKey: '...'
});

let client = Client.fromAuthenticationProvider(authProvider, Mqtt);

client.setRetryPolicy(new NoRetry());
client.open((err, result) => {

    if(err)
        console.log('client open error: %s', err);

        client.on('error', err => {
        console.log('client on error called.');
    });

    client.on('disconnect', () => {
        console.log('client on disconnect called.');
    });

    sendMessage();

    setInterval(() => {
        sendMessage();
    }, 10000);

});

let i = 0;

function sendMessage() {
    let sendAsync = Promise.promisify(client.sendEvent, { context: client });
    let message = new Message(data);
    let messageI = ++i;
    console.log('SEND: ', messageI);
    sendAsync(message).then(result => {
        console.log('CB: ', messageI);
    }, err => {
        console.log('CB with err: ', messageI);
        console.error(err);
    }).timeout(5000)
    .catch(Promise.TimeoutError, err => {
        console.log('TIMEOUT', messageI);
        console.log('TRANSPORT STATE: ', client._transport._fsm.state);
        console.log('TRANSPORT MQTT STATE: ', client._transport._mqtt._fsm.state);

        // WORKAROUND
        //   On raspbian, the socket is never closed. 'close' is never called.
        //   Hit the state machine manually to avoid staying 'connected' forever.
        if(client._transport._mqtt._fsm.state == 'connected')
            client._transport._mqtt._fsm.handle('closeEvent');
    });
}
```

# Console log of the issue:
_(without workaround timeout)_
```
Debugger listening on ws://0.0.0.0:9229/b10ed993-32e0-410a-8766-4ec2e6ae76ac
For help see https://nodejs.org/en/docs/inspector
Debugger attached.
  azure-iot-device-mqtt:Mqtt disconnected -> connecting (disconnected.connect) +0ms
  azure-iot-device-mqtt:Mqtt topic publish: devices/dev-slm-client/messages/events/ +1ms
  azure-iot-device-mqtt:Mqtt topic subscribe: devices/dev-slm-client/messages/devicebound/# +1ms
  azure-iot-mqtt-base:MqttBase disconnected -> connecting (disconnected.connect) +0ms
  azure-iot-mqtt-base:MqttBase username: <...>.azure-devices.net/dev-slm-client/DeviceClientType=azure-iot-device%2F1.3.0&api-version=2017-06-30 +0ms
  azure-iot-mqtt-base:MqttBase uri:      mqtts://<...>.azure-devices.net +1ms
(node:4307) [INSPECTOR_ASYNC_STACK_TRACES_NOT_AVAILABLE] Warning: Warning: Async stack traces in debugger are not available on 32bit platforms. The feature is disabled.
  azure-iot-mqtt-base:MqttBase Device is connected +409ms
  azure-iot-mqtt-base:MqttBase CONNACK: {"cmd":"connack","retain":false,"qos":0,"dup":false,"length":2,"topic":null,"payload":null,"sessionPresent":true,"returnCode":0} +0ms
  azure-iot-mqtt-base:MqttBase connecting -> connected () +1ms
  azure-iot-device-mqtt:Mqtt connect +412ms
  azure-iot-device-mqtt:Mqtt connecting -> connected () +0ms
SEND:  1
  azure-iot-device-mqtt:Mqtt sendEvent {"data":...}","properties":{"propertyList":[]},"messageId":"","to":"","lockToken":"","correlationId":"","userId":""} +3ms
  azure-iot-device-mqtt:Mqtt PUBACK: {"transportObj":{"cmd":"publish","topic":"devices/dev-slm-client/messages/events/","payload":"{...}","qos":1,"retain":false,"messageId":37302,"dup":false}} +8s
CB:  1
======> NETWORK CABLE UNPLUGGED HERE <========
SEND:  2
  azure-iot-device-mqtt:Mqtt sendEvent {"data":...}","properties":{"propertyList":[]},"messageId":"","to":"","lockToken":"","correlationId":"","userId":""} +10s
SEND:  3
  azure-iot-device-mqtt:Mqtt sendEvent {"data":...}","properties":{"propertyList":[]},"messageId":"","to":"","lockToken":"","correlationId":"","userId":""} +1m
SEND:  4
  azure-iot-device-mqtt:Mqtt sendEvent {"data":...}","properties":{"propertyList":[]},"messageId":"","to":"","lockToken":"","correlationId":"","userId":""} +1m
SEND:  5
  azure-iot-device-mqtt:Mqtt sendEvent {"data":...}","properties":{"propertyList":[]},"messageId":"","to":"","lockToken":"","correlationId":"","userId":""} +1m
```

# Final thoughts

Could be the same or related to #175, but i'm not completly shure.
Could also be the case when using AMQP. I switched to AMQP and back early when investigating with same results.
... and could also be a problem of the underlying libs (mqtt, end-of-stream, node itself), but maybe it can be caught on api-level.

**Comment by pierreca** [comment]
Hi @sgacond, thanks for filing this issue.
what is surprising is that [we are setting a keepalive value](https://github.com/Azure/azure-iot-sdk-node/blob/master/common/transport/mqtt/src/mqtt_base.ts#L216) that sends a ping every 3 minutes - so at worst after 6 minutes the MQTT stack should tell the Azure IoT SDK that the connection is unresponsive, even when the socket is not killed. I've tested this to work on Ubuntu, I wonder if there's any difference with Raspian (shouldn't be...). I'll need to setup a device to check that.

I'm not declaring this a bug yet until we find the root cause but there's definitely something the SDK should be doing for you here. if you keep digging and find new information please do keep us posted, and we'll do the same as we investigate. Thanks!

**Comment by sgacond** [comment]
Hi @pierreca 
If I remember correctly, the ping was always rescheduled when trying to send e message, but i'm not sure. But I kept my Test-setup as it is to continue and test again. 
So I'm going to run another test to confirm this. The expected behaviour should be that the after 3 minutes, the state (client._transport._mqtt._fsm.state) should shift to disconnected in any case? So with and without sending events in between?
I hope I find some time to do this within the next few days.

**Comment by pierreca** [comment]
@sgacond pings should not be rescheduled (see https://github.com/Azure/azure-iot-sdk-node/blob/master/common/transport/mqtt/src/mqtt_base.ts#L217)

the disconnect should take anywhere between 3 to 6 minutes to fire, the reason is that the way the mqtt library is written, the ping response of time T is verified at T+1 (in other words, in the worst case, at time T, the device is connected and the ping comes back immediately, then gets disconnected - the response gets verified at T+1 and is OK and a new ping request is sent while disconnected - then at T+2 the device realizes that the T+1 response was never received.

**Comment by pierreca** [comment]
@sgacond did you end up running another test?

**Comment by sgacond** [comment]
Hi @pierreca 
First of all, sorry for the long delay!
I did run some more tests but was not able to reproduce the issue anymore. As you wrote above, after twice the timeout time (6 mins), the socket is reset correctly. I did longer running tests in the past and had the behaviour of not resetting, but I don't get it reproduced anymore.
Ok to close for me... I'll write again, when i'm able to reproduce again.
Thank you very much!
